rmk/src/split/serial/mod.rs

```diff
@@ -42,44 +43,90 @@ pub(crate) async fn run_serial_peripheral_monitor<
 /// Serial driver for BOTH split central and peripheral
 pub(crate) struct SerialSplitDriver<S: Read + Write> {
     serial: S,
+    buffer: [u8; SPLIT_MESSAGE_MAX_SIZE],
+    n_bytes_part: usize,
 }
 
 impl<S: Read + Write> SerialSplitDriver<S> {
     pub(crate) fn new(serial: S) -> Self {
-        Self { serial }
+        Self {
+            serial,
+            buffer: [0_u8; SPLIT_MESSAGE_MAX_SIZE],
+            n_bytes_part: 0,
+        }
     }
 }
 
 impl<S: Read + Write> SplitReader for SerialSplitDriver<S> {
-    async fn read(&mut self) -> Result<SplitMessage, SplitDriverError> {
-        let mut buf = [0_u8; SPLIT_MESSAGE_MAX_SIZE];
-        let n_bytes = self
-            .serial
-            .read(&mut buf)
-            .await
-            .map_err(|_e| SplitDriverError::SerialError)?;
-        if n_bytes == 0 {
-            return Err(SplitDriverError::EmptyMessage);
+    async fn read(&mut self) -> Result<Vec<SplitMessage, 2>, SplitDriverError> {
+        const SENTINEL: u8 = 0x00;
+        let mut messages = Vec::new();
+        while self.n_bytes_part < self.buffer.len() {
+            let n_bytes = self
+                .serial
+                .read(&mut self.buffer[self.n_bytes_part..])
+                .await
+                .inspect_err(|_e| self.n_bytes_part = 0)
+                .map_err(|_e| SplitDriverError::SerialError)?;
+            if n_bytes == 0 {
+                return Err(SplitDriverError::EmptyMessage);
+            }
+
+            self.n_bytes_part += n_bytes;
+            if self.buffer[..self.n_bytes_part].contains(&SENTINEL) {
+                break;
+            }
         }
-        let message: SplitMessage = postcard::from_bytes(&buf).map_err(|e| {
-            error!("Postcard deserialize split message error: {}", e);
-            SplitDriverError::DeserializeError
-        })?;
-        Ok(message)
+
+        let mut start_byte = 0;
+        let mut end_byte = start_byte;
+        while end_byte < self.n_bytes_part {
```

Comment:
The buffer size is `SPLIT_MESSAGE_MAX_SIZE`, so this second loop is not needed -- we can assume that after the first `while` loop, there's a complete `SplitMessage` received.